readthedocs/oauth/services/gitlab.py

```diff
@@ -0,0 +1,235 @@
+"""OAuth utility functions"""
+
+import logging
+import json
+import re
+
+from django.conf import settings
+from requests.exceptions import RequestException
+from allauth.socialaccount.models import SocialToken
+from allauth.socialaccount.providers.gitlab.views import GitLabOAuth2Adapter
+from urlparse import urljoin
+
+from readthedocs.restapi.client import api
+
+from ..models import RemoteOrganization, RemoteRepository
+from .base import Service, DEFAULT_PRIVACY_LEVEL
+
+
+log = logging.getLogger(__name__)
+
+
+class GitLabService(Service):
+    """Provider service for GitLab"""
+
+    adapter = GitLabOAuth2Adapter
+    url_pattern = re.compile(re.escape(adapter.provider_base_url))
+    default_avatar = {
+        'repo': urljoin(settings.MEDIA_URL, 'images/fa-bookmark.svg'),
+        'org':  urljoin(settings.MEDIA_URL, 'images/fa-users.svg'),
+    }
+
+    def paginate(self, url, **kwargs):
+        """Combines return from GitLab pagination. GitLab uses
+        LinkHeaders, see: http://www.w3.org/wiki/LinkHeader
+
+        :param url: start url to get the data from.
+        :param kwargs: optional parameters passed to .get() method
+
+        See http://doc.gitlab.com/ce/api/README.html#pagination
+        """
+        resp = self.get_session().get(url, data=kwargs)
+        result = resp.json()
+        next_url = resp.links.get('next', {}).get('url')
+        if next_url:
+            result.extend(self.paginate(next_url, **kwargs))
+        return result
+
+    def sync(self):
+        """Sync repositories from GitLab API"""
+        org = None
+        repos = self.paginate(
+            u'{url}/api/v3/projects'.format(url=self.adapter.provider_base_url),
+            per_page=100,
+            order_by='path',
+            sort='asc'
+        )
+        for repo in repos:
+            # Skip archived repositories
+            if repo.get('archived', False):
+                continue
+            if not org or org.slug != repo['namespace']['id']:
+                org = self.create_organization(repo['namespace'])
+
+            self.create_repository(repo, organization=org)
+
+    def create_repository(self, fields, privacy=DEFAULT_PRIVACY_LEVEL,
+                          organization=None):
+        """Update or create a repository from GitLab API response
+
+        :param fields: dictionary of response data from API
+        :param privacy: privacy level to support
+        :param organization: remote organization to associate with
```

Comment:
@connorshea, @destroyerofbuilds of course you're both right, but i had to adapt the given api and this api uses `organization`. My recommendation would be to get this into master and then discuss how to refactor the oauth-plugin system.